Batch correction tutorial for single cell

MarisaJL opened the pull request:
Batch Correction/Integration tutorial for Scanpy and Seurat

TODO

- [x] Hands on sections for Scanpy
- [x] Add more background/explanation on what batch correction/integration does
- [x] Write last section comparing before and after batch correction

MarisaJL committed:
Create batch correction tutorial

MarisaJL committed:
Merge branch 'batch_correct' of https://github.com/MarisaJL/training-material into batch_correct

MarisaJL committed:
Add images

MarisaJL committed:
Minor edits

MarisaJL committed:
Fixing citation

MarisaJL committed:
Fixing zenodo link

MarisaJL committed:
Trying to fix error

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+
+If we simply run these different batches or combined datasets through a clustering pipeline such as Scanpy or Seurat, we might not get useful results. Clustering prioritises the genes that show the biggest differences in expression and uses these to identify groups of cells that share similar expression patterns. When we have different experimental batches or have combined multiple studies, these ...
+
+In order to look beyond these technical differences, we can perform batch correction or integration. Both the Scanpy and Seurat pipelines include tools that can be used to correct for differences between experimental batches or to integrate datasets - we actually use the same tools to do both. In this tutorial, we will learn how to use these tools in either the Scanpy or Seurat pipeline - you can ...
```
```suggestion
In order to look beyond these technical differences, we can perform batch correction or integration. Both the Scanpy and Seurat pipelines include tools that can be used to correct for differences between experimental batches or to integrate datasets - we actually use the same tools to do both. In this tutorial, you will learn how to use these tools in either the Scanpy or Seurat pipeline - you can choose which one you would like to use.
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+
+# Scanpy or Seurat?
+
+Scanpy and Seurat are two of the most commonly used pipelines (sets of tools) for analysing single cell data. Both pipelines have all the tools required to perform clustering, which identifies groups of cells that share similar expression profiles. Clustering is often the first step in single cell analysis because it makes our data easier to interpret. Clusters represent groups of cells that are  ...
```
```suggestion
Scanpy and Seurat are two of the most commonly used pipelines (sets of tools) for analysing single cell data. Both pipelines have all the tools required to perform clustering, which identifies groups of cells that share similar expression profiles. Clustering is often a key goal in single cell analysis because it makes our data easier to interpret. Clusters represent groups of cells that are expressing the same genes, which often correspond to specific cell types or states. Our goal is to identify biologically relevant clusters that will help us to better understand our data.
```

nomadscientist commented:
Kind of a general comment that there's a lot of switching between 'we' and 'you' forms which is a bit jarring, but relatively minor!

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+> > <solution-title></solution-title>
+> >
+> > 1. The dataset that we're using comes from a study that compared different single cell techniques. The `Method` column tells us which technique was used on each cell.
+> > 2. Each experimental technique can be considered as its own experimental batch. Each of these batches was processed independently, which by itself can be enough to require batch correction, even if the same experimental protocol is used - batches can vary simply because they were processed at different times or by different people in the same lab! In this case, we have an even stronger reason ...
```
```suggestion
> > 2. Each experimental technique can be considered as its own experimental batch. Each of these batches was processed independently, which by itself can be enough to require batch correction, even if the same experimental protocol is used. Batches can vary simply because they were processed at different times or by different people in the same lab! In this case, we have an even stronger reason to believe that these batches will differ - we know that each batch was produced using a different technique. It seems likely that we'll need to perform batch correction. We would consider this to be batch correction rather than integration because these data all came from the same original study.
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+> 1. {% tool [Seurat Data Management](toolshed.g2.bx.psu.edu/repos/iuc/seurat_data/seurat_data/5.0+galaxy0) %} with the following parameters:
+>    - *"Method used"*: `Inspect Seurat Object`
+>        - *"Display information about"*: `Cell Metadata`
+>
```
<img width="298" alt="Screenshot 2025-02-13 at 15 16 14" src="https://github.com/user-attachments/assets/e5af0953-b9c3-4102-95e8-a8ee476e2da9" />

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+
+> <comment-title></comment-title>
+>
+> The cell metadata is any information about the cells that the original authors have included in the SeuratObject. As well as the cell barcode or identifier for each individual cell, the metadata will usually include information such as which donor or sample the cell came from or which experimental group it was in. Sometimes, this metadata will include lots of useful details such as demographic  ...
```
```suggestion
> The cell metadata is any information about the cells that the original authors have included in the SeuratObject. As well as the cell barcode or identifier for each individual cell, the metadata will usually include information such as which donor or sample the cell came from, or which experimental group it was in. Sometimes, this metadata will include lots of useful details, such as demographic information about human donors. Such information can help us to better understand our results. 
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+
+The terms batch correction and integration can be used somewhat interchangably, because they both refer to the same process of looking for shared cell subpopulations across groups. The same tools are used in the same way for both procedures, so you could use the workflow described in this tutorial to perform integration as well as batch correction.
+
+The only difference is that we tend to talk about batch correction when we are working with groups produced in a single study (e.g. different experimental batches), while we would say integration when we're combining separate datasets from multiple studies.
```
This last point is SUPER important - which is which? Perhaps move earlier, or change the title of this section to "Batch correction or integration?"

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+
+Splitting the batches into separate layers could help to address some of the technical differences between them because of the separate preprocessing, but we'll have to wait for the results to see if this has been enough to eliminate these differences.
+
+If you want to understand the impact of splitting and preprocessing the batches separately, then you could skip ahead to the next section, **Clustering with Seurat** and compare your results to those shown in this tutorial.
```
I didn't quite understand this. What did you mean?

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+> > <solution-title></solution-title>
+> >
+> > 1. We can see that there are now 9 layers in our SeuratObject. 
+> > 2. We started out with one layer of raw data, called `counts`. That layer has now been split up according to `Method`. We now have nine `counts` layers. Each layer represents one of the batches named in the `Method` column of the cell metadata. We can see the names of the methods in the layer names. For example, the counts.Drop-seq layer contains the raw counts produced using the Drop-seq tec ...
```
```suggestion
> > 2. We started out with one layer of raw data, called `counts`. That layer has now been split up according to `Method`. We now have nine `counts` layers. Each layer represents one of the batches named in the `Method` column of the cell metadata. We can see the names of the methods in the layer names. For example, the `counts.Drop-seq` layer contains the raw counts produced using the Drop-seq technique. Seven different methods were used in this study, but one of them was applied to three different batches - you should be able to see three layers with `Chromium_v2` in their names.
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+We'll follow the default Seurat pipeline here, except that we'll use `30` PCs to build the neighborhood graph and cluster with a resolution of `2` as these were the parameters used in [the original Seurat version of this tutorial](https://satijalab.org/seurat/articles/seurat5_integration). We'll also give our clusters and UMAP more recognisable names as we'll be running these tools again later, a ...
+
+> <comment-title></comment-title>
+> Seurat has another option for preprocessing - rather than use the three separate functions presented below, you can use a single function called `SCTransform` to preform normalisation, identification of variable genes, and scaling all in one go. You will find this option on Galaxy's {% tool Seurat Preprocessing} tool.
```
```suggestion
> Seurat has another option for preprocessing - rather than use the three separate functions presented below, you can use a single function called `SCTransform` to preform normalisation, identification of variable genes, and scaling all in one go. You will find this option on Galaxy's {% tool Seurat Preprocessing %} tool.
```

nomadscientist replied:
```suggestion
> Seurat has another option for preprocessing - rather than use the three separate functions presented below, you can use a single function called `SCTransform` to preform normalisation, identification of variable genes, and scaling all in one go. You will find this option on Galaxy's {% icon tool %} **Seurat Preprocessing** tool.
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+> <comment-title></comment-title>
+> Seurat has another option for preprocessing - rather than use the three separate functions presented below, you can use a single function called `SCTransform` to preform normalisation, identification of variable genes, and scaling all in one go. You will find this option on Galaxy's {% tool Seurat Preprocessing} tool.
+>
+> If you use `SCTransform` for preprocessing then you'll need to click the button to choose `Yes` for `Use SCT as Normalization Method` when you run `IntegrateLayers`. The `SCTransform` normalises the data in its own way, so we just need to let the tool know what to expect!
```
```suggestion
> If you use `SCTransform` for preprocessing, then you'll need to choose `Yes` for `Use SCT as Normalization Method` when you run `IntegrateLayers`. The `SCTransform` normalises the data in its own way, so we just need to let the tool know what to expect!
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+>
+> If you use `SCTransform` for preprocessing then you'll need to click the button to choose `Yes` for `Use SCT as Normalization Method` when you run `IntegrateLayers`. The `SCTransform` normalises the data in its own way, so we just need to let the tool know what to expect!
+>
+> The next step after identifying clusters would usually be to look for marker genes that are differentially expressed between clusters. If you perform integration/batch correction after using `SCTransform` then you will need to run the `PrepSCTFindMarkers` function before using tools such as `FindMarkers`. You'll find this in the {% tool Seurat Integrate %} tool.
```
```suggestion
> The next step after identifying clusters would usually be to look for marker genes that are differentially expressed between clusters. If you perform integration/batch correction after using `SCTransform`, then you will need to run the `PrepSCTFindMarkers` function before using tools such as `FindMarkers`. You'll find this in the {% tool Seurat Integrate %} tool.
```

nomadscientist replied:
```suggestion
> The next step after identifying clusters would usually be to look for marker genes that are differentially expressed between clusters. If you perform integration/batch correction after using `SCTransform` then you will need to run the `PrepSCTFindMarkers` function before using tools such as `FindMarkers`. You'll find this in the {% icon tool %} **Seurat Integrate** tool.
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+>    >
+>    > We will use the output from `RunPCA` in the  following section when we perform batch correction.
+>    >
+>    > If you're already very familiar with the Seurat clustering pipeline and you just want to try using the  {% tool Seurat Integrate %} tools, then you can skip ahead to the **Clustering after Integration** step now.
```
```suggestion
>    > If you're already familiar with the Seurat clustering pipeline and you just want to try using the  {% icon tool %} **Seurat Integrate** tools, then you can skip ahead to the **Clustering after Integration** step now.
```

nomadscientist replied:
Do these 'skips' work for running through the tutorial? Or is that assuming they won't actually be using Galaxy while reading the tutorial?

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+>        - *"Algorithm for modularity optimization"*: `1. Original Louvain`
+>        - *"Name for output clusters"*: `unintegrated_clusters`
+>
+>    > <comment-title> short description </comment-title>
```
```suggestion
>    > <warning-title> short description <warning-title>
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+>    > <comment-title> short description </comment-title>
+>    >
+>    > Make sure that you change the default name for the clusters to `unintegrated_clusters`! 
+>    {: .comment}
```
```suggestion
>    {: .warning}
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+>        - In *"Advanced Options"*:
+>            - *"Name for dimensional reduction"*: `umap.unintegrated`
+>
+>    > <comment-title> short description </comment-title>
```
```suggestion
>    > <warning-title> short description </warning-title>
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+>    > <comment-title> short description </comment-title>
+>    >
+>    > Make sure that you change the default name for the UMAP results to `umap.unintegrated`! 
+>    {: .comment}
```
```suggestion
>    {: .warning}
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+
+# Clustering without Batch Correction
+
+We suspect that batch correction will be needed because of the different technologies used to construct this dataset, but we'll try clustering without any correction first. This will confirm whether batch correction is truly needed on the basis of `Method`. Comparing the results we get now with those we'll get after batch correction should also help us to understand what batch correction is doing ...
```
Maybe something more like "First, we need to determine whether we even need to perform batch correction. Therefore, we will try clustering without correction first...."

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+
+> <comment-title> short description </comment-title>
+>
+> {% tool Seurat Integrate %} provides several integration methods, which all perform the integration or batch correction in their own way. You might want to experiment by using one of the other methods to see how it affects the results. When you are working on your own data, it can be a good idea to try a few different integration methods to see which one produces the best results. The best inte ...
```
```suggestion
> {% tool Seurat Integrate %} provides several integration methods, which all perform the integration or batch correction in their own way. You might want to experiment by using different methods to see how it affects the results. When you are working on your own data, it can be a good idea to try a few different integration methods to see which one produces the best results. The best integration or batch correction would be the one that eliminates the most of the technical differences between datasets or batches while producing biologically meaningful results. If we end up with completely unexpected results rather than clusters that match up well with known cell types, then we know that something has gone wrong!
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+>        - *"Integration method to use"*: `CCA Integration`
+>        - *"Name for new dimensional reduction"*: `integrated.cca`
+>
+>    > <comment-title> short description </comment-title>
```
```suggestion
>    > <comment-title> Remember the name </comment-title>
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+>
+{: .hands_on}
+
+It's good practice to rejoin our layers now, so that those separate layers or batches will end up back in the same layer. We don't actually need to do this now as it won't affect the clustering results, but it is important if we want to perform downstream analyses such as Differential Expression analysis.
```
```suggestion
It's good practice to rejoin our layers now, so that those separate layers/batches will end up together. We don't actually need to do this now (as it won't affect the clustering results), but it is important if we want to perform downstream analyses such as Differential Expression analysis.
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+> > <solution-title></solution-title>
+> >
+> > 1. The first plot shows 25 clusters (remember that Seurat starts from cluster 0!). Although the high resolution means we still have plenty of clusters, the batch correction has reduced the number. The clusters also look less fragmented than they did before.
+> > 2. When we colour in the plot by `Method` we can see that all the colours are mixed together across all of the clusters. We don't have any clusters that are all one colour and there aren't any big patches of colour. The batch correction has successfully removed the differences between the batches so that they're no longer dominating the results.
```
```suggestion
> > 2. When we colour in the plot by `Method`, we can see that all the colours are mixed together across all of the clusters. We don't have any clusters that contain only one colour. The batch correction has successfully removed the differences between the batches so that they're no longer dominating the results.
```

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+
+
+
+> <comment-title></comment-title>
```
I think this is quite important, so probably not a comment. Could show the steps via bullets, i.e. "You would then...."

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+>
+>If you look back at the cell metadata table we created at the beginning of this tutorial, you'll see there is an annotation called `CellType`. We can colour in our UMAPs using this annotation instead of the `Method`. If our clusters make biological cell sense, we should see that these cell types are clumped together because cells of the same type should be close to each other.
+>
+> If the cell types are all blended together across the entire UMAP (as with our `Method` plot) then this would be a sign that something has gone wrong. When we are performing batch correction or integration, there is a risk that we could over-integrate the data, eliminating the biological differences we're interested in alongside the technical differences we wanted to remove. 
```
For training purposes, it would be really valuable to give an example of over-integrated data - you don't have to make them calculate it, but adding images as a comment or detail or something else would be really helpful.

nomadscientist commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,529 @@
+
+In this tutorial, we've learned how to perform batch correction or integration when analysing single cell data with either the Scanpy or Seurat pipelines. If you want to learn more about these pipelines then you might want to try analysing a slightly trickier dataset in the [Scanpy]({% link topics/single-cell/tutorials/scrna-case_basic-pipeline/tutorial.md %}) or [Seurat]({% link topics/single-ce ...
+
+This tutorial is part of the https://singlecell.usegalaxy.eu portal ({% cite tekman2020single %}).
```
I would remove this line about the data portal, personally.

MarisaJL committed:
Merge branch 'galaxyproject:main' into batch_correct

MarisaJL committed:
Apply suggestions from code review

Co-authored-by: Wendi Bacon <44605769+nomadscientist@users.noreply.github.com>

MarisaJL committed:
Revise tutorial for clarity and authorship details

Updated tutorial to clarify authorship and editing roles, and improved explanations on batch correction and integration processes.

MarisaJL committed:
Add integration images for Scanpy

MarisaJL committed:
Add workflow for Scanpy

MarisaJL committed:
Update input/output labels in Scanpy workflow

MarisaJL committed:
Add Scanpy test workflow for single-cell analysis

MarisaJL committed:
Add hands on sections for Scanpy

MarisaJL committed:
Rename workflow_Scanpy-test.yml to workflow_Scanpy-tests.yml

MarisaJL committed:
Rename workflow_Seurat-test.yml to workflow_Seurat-tests.yml

MarisaJL committed:
Write Scanpy sections and questions

MarisaJL committed:
Convert file line endings from CRLF to LF

MarisaJL committed:
Fix error for Inspect_before_Split output

MarisaJL committed:
Revise batch correction tutorial for clarity and updates

Updated tutorial on batch correction and integration with Seurat and Scanpy, including new links and clarifications.

MarisaJL committed:
Update data-library.yaml with new DOI and URLs

MarisaJL committed:
Update Scanpy images

MarisaJL committed:
Delete old image

MarisaJL committed:
Delete old image

MarisaJL committed:
Correct image file names

MarisaJL committed:
Add CellType pictures

MarisaJL committed:
Add final section comparing before and after plots

MarisaJL committed:
Minor edits

MarisaJL committed:
Add before and after pics

MarisaJL committed:
Update Zenodo links and correct indentation

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,784 @@
+
+Single cell analyses can be complex. We may have data from different experimental batches, perhaps because we ran our experiments at different times, in different labs, or using different sequencing platforms. Sometimes we might want to combine multiple datasets, for example if we want to compare our own experimental data to a similar public dataset.
+
+If we simply run these different batches or combined datasets through a clustering pipeline such as Scanpy or Seurat, we might not get useful results. Clustering prioritises the genes that show the biggest differences in expression and uses these to identify groups of cells that share similar expression patterns. When we have different experimental batches or have combined multiple studies, these ...
```
```suggestion
Running different batches or combined datasets through a clustering pipeline like Scanpy or Seurat without any pre-processing may not yield useful results. This is because clustering works by identifying genes with the largest differences in expression and grouping cells that share similar expression patterns. When data comes from multiple experimental batches or studies, however, the largest sources of variation often reflect technical differences between batches rather than the biology of interest like cell type. As a result, clusters may end up capturing batch or dataset identity rather than anything biologically meaningful.
```

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,784 @@
+
+If we simply run these different batches or combined datasets through a clustering pipeline such as Scanpy or Seurat, we might not get useful results. Clustering prioritises the genes that show the biggest differences in expression and uses these to identify groups of cells that share similar expression patterns. When we have different experimental batches or have combined multiple studies, these ...
+
+In order to look beyond these technical differences, we can perform batch correction or integration. Both the Scanpy and Seurat pipelines include tools that can be used to correct for differences between experimental batches or to integrate datasets - we actually use the same tools to do both. In this tutorial, you will learn how to use these tools in either the Scanpy or Seurat pipeline - you ca ...
```
```suggestion
To look beyond these technical differences, we can perform batch correction or integration. Both Scanpy and Seurat include tools for correcting differences between experimental batches and integrating datasets, and in practice we use the same tools for both. In this tutorial, you will learn how to use these tools in either pipeline. The choice of Scanpy or Seurat is yours.
```

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,784 @@
+
+# Batch Correction or Integration?
+
+We will often need to perform batch correction or integration during single cell analyses. If we are working with different experimental batches, donors, conditions, or datasets, then we need to look beyond the technical differences between them. Batch correction or integration may be required to enable us to do this. These techniques work by matching cells of similar types or states across batch ...
```
```suggestion
When working with multiple experimental batches, donors, conditions, or datasets, we need to look beyond the technical differences between them and batch correction or integration are the techniques we use to do this. Both work by identifying cell subpopulations that are shared across groups, effectively matching cells of similar types or states.
```

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,784 @@
+
+We will often need to perform batch correction or integration during single cell analyses. If we are working with different experimental batches, donors, conditions, or datasets, then we need to look beyond the technical differences between them. Batch correction or integration may be required to enable us to do this. These techniques work by matching cells of similar types or states across batch ...
+
+The terms batch correction and integration are closely related and you may see them being used somewhat interchangably, because they both refer to the same process of looking for shared cell subpopulations across groups. The same tools are used in the same way for both procedures, so you could use the workflow described in this tutorial to perform integration as well as batch correction.
```
```suggestion
The terms batch correction and integration are closely related and are often used interchangeably, since they refer to the same underlying process and use the same tools in the same way. The distinction is mainly one of context: batch correction typically refers to removing unwanted technical variation between groups from a single study, such as different experimental batches, while integration refers to aligning cell populations across separate datasets from multiple studies.
```

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,784 @@
+
+The terms batch correction and integration are closely related and you may see them being used somewhat interchangably, because they both refer to the same process of looking for shared cell subpopulations across groups. The same tools are used in the same way for both procedures, so you could use the workflow described in this tutorial to perform integration as well as batch correction.
+
+The only difference is that we tend to talk about batch correction when we are working with groups produced in a single study (e.g. different experimental batches), while we would say integration when we're combining separate datasets from multiple studies. Batch correction is performed to remove unwanted technical variation, while integration aims to align cell populations across studies.
```
I suggested merging the two paragraphs in the middle as one, so If my suggestions works for you, we do not need this paragraph.

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,784 @@
+
+# Scanpy or Seurat?
+
+Scanpy and Seurat are two of the most commonly used pipelines (sets of tools) for analysing single cell data. Both pipelines have all the tools required to perform clustering, which identifies groups of cells that share similar expression profiles. Clustering is often a key goal in single cell analysis because it makes our data easier to interpret. Clusters represent groups of cells that are expr ...
```
```suggestion
Scanpy and Seurat are two of the most commonly used pipelines for analysing single cell data. Both include tools for preprocessing, dimensional reduction (such as PCA), neighbourhood graph construction, and clustering. Clustering is often a key goal in single cell analysis, as it groups cells with similar expression profiles. These groups frequently correspond to specific cell types or states, making the data easier to interpret and understand.
```

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,784 @@
+
+Scanpy and Seurat are two of the most commonly used pipelines (sets of tools) for analysing single cell data. Both pipelines have all the tools required to perform clustering, which identifies groups of cells that share similar expression profiles. Clustering is often a key goal in single cell analysis because it makes our data easier to interpret. Clusters represent groups of cells that are expr ...
+
+The Scanpy and Seurat pipelines include tools for preprocessing single cell data, performing dimensional reductions such as PCA, constructing a neighbourhood graph and finding clusters in it. Although both pipelines perform the same basic steps, there are some differences in how these steps are performed that mean you can end up with slightly different results depending on which pipeline you choo ...
```
```suggestion
Although both pipelines follow the same basic steps, small differences in how those steps are implemented mean results can vary slightly depending on your choice. Broadly speaking, though, you should reach the same conclusions either way.
```

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,784 @@
+
+Let's take a look at our data before we begin the analysis to see whether we might need to perform batch correction or integration. While batches or combined datasets often do require correction, we should examine the data first to check whether it is necessary.
+
+Correcting batch effects is important, but we also have to be careful about overcorrecting or overintegrating our data. If we apply batch correction or integration tools when they aren't needed, we could end up removing the real biological differences we're interested in. Even when some correction is needed, we should be careful to avoid overcorrecting and losing these biological differences. We' ...
```
```suggestion
Correcting batch effects is important, but we also need to be careful not to overcorrect. Applying batch correction or integration when it isn't needed, or overcorrecting when it is, risks removing the very biological differences we are interested in. Either way, it is important to interpret results carefully to determine whether they reflect true biological variation or technical artefacts.
```

MarisaJL committed:
Apply suggestions from code review

Co-authored-by: Diana Chiang Jurado <122611454+dianichj@users.noreply.github.com>

MarisaJL committed:
Revise tutorial content on batch correction and integration

Reworded sections for clarity and improved flow regarding batch correction and integration in single-cell analyses.

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,783 @@
+
+> <comment-title></comment-title>
+>
+> The cell metadata is any information about the cells that the original authors have included with the dataset. As well as the cell barcode or identifier for each individual cell, the metadata will usually include information such as which donor or sample the cell came from, or which experimental group it was in. Sometimes, this metadata will include lots of useful details, such as demographic i ...
```
```suggestion
> The cell metadata is any information about the cells that the original authors have included with the dataset. As well as the cell barcode or identifier for each individual cell, the metadata will usually include information such as which donor or sample the cell came from, or which experimental group it was in. Sometimes, this metadata will include a lot of useful details, such as demographic information about human donors. This information can help us to better understand our results.
```

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,783 @@
+>    >
+>    > We will use the output from `Scanpy normalize` in the  following section when we perform batch correction.
+>    >
+>    > If you're already familiar with the Scanpy clustering pipeline and you just want to try using the  {% icon tool %} **Scanpy remove confounders** tools, then you can skip ahead to the **Clustering after Integration** step now. In a real analysis, it would be best to complete the clustering without batch correction first to check if it is needed.
```
I can't remember if this was mentioned earlier - it would be best to complete the clustering without batch correction first to check if it is needed. Perhaps, this can be suggested before performing all the steps. Maybe at line 217 when the user takes a closer look at whether the data might need batch correction.

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,783 @@
+Let's take a look at our data before we begin the analysis to see whether we might need to perform batch correction or integration. While batches or combined datasets often do require correction, we should examine the data first to check whether it is necessary.
+
+Correcting batch effects is important, but we also need to be careful not to overcorrect. Applying batch correction or integration when it isn't needed, or overcorrecting when it is, risks removing the very biological differences we are interested in. Either way, it is important to interpret results carefully to determine whether they reflect true biological variation or technical artefacts.
+
```
Even though this tutorial is likely for advanced users, it will be useful to mention at which step of the single-cell analysis pipeline batch correction should be considered or performed. For example, do we perform batch correction during the pre-processing step? Before or after normalization? . Just leaving my comment here, you will likely have a better idea on where to write about it :)

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,783 @@
+>
+> > <solution-title></solution-title>
+> >
+> > 1. The first plot is coloured by cluster. We can see there are 47 clusters (Scanpy numbers the clusters starting from 0). That's a lot - this is partly because we used a relatively high clustering resolution, but these fragmented clusters could also be a sign that something has gone wrong with our analysis.
```
I like this content. I didn't know we could play a bit with resolutions to see if there are batch effects. Perhaps this could be emphasized if you consider it is a good idea.

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,783 @@
+
+In the Seurat pipeline, we can use the ability of the SeuratObject to store multiple layers to split our data up before we begin the analysis. We might do this when we have different batches or if we've combined datasets.
+
+Splitting our data into layers means that the Seurat preprocessing tools can work on each layer separately. Seurat can treat each layer as if it were a separate dataset during preprocessing. Each layer (in this case, each of our batches) will be normalised independently. We'll also identify the highly variable genes within each batch, rather than across the whole dataset. Seurat will then create  ...
```
This layers part, I do not understand it so well. Is splitting layers the same as separating samples and normalizing each sample individually? Is layers a term that means different in Seurat context?

MarisaJL replied:
Yes, its a Seurat thing - you separate each batch into a layer within the same dataset, so it is a bit like separating samples but it's something their integration tools expect you to have done

dianichj replied:
Thanks for the explanation :)!

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,783 @@
+>
+> The next step after identifying clusters would usually be to look for marker genes that are differentially expressed between clusters. If you perform integration/batch correction after using `SCTransform`, then you will need to run the `PrepSCTFindMarkers` function before using tools such as `FindMarkers`. You'll find this in the {% tool Seurat Integrate %} tool.
+>
+> The rest of the workflow will be the same as shown in this tutorial, but you will end up with different results because `SCTransform` handles preprocessing in a slightly different way than the three separate tools. If you want to learn more about these differences then you can choose the SCTransform route in the [Clustering 3k PBMCs with Seurat]({% link topics/single-cell/tutorials/scrna-seurat ...
```
This layers section is very important but it is a bit confusing. The steps are not so clear to me from line 399 to line 403

dianichj replied:
I see it is a comment not a hands-on but it is hard to understand for me.

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,783 @@
+>    >
+>    > We will use the output from `RunPCA` in the  following section when we perform batch correction.
+>    >
+>    > If you're already familiar with the Seurat clustering pipeline and you just want to try using the  {% icon tool %} **Seurat Integrate** tools, then you can skip ahead to the **Clustering after Integration** step now. In a real analysis, it would be best to complete the clustering without batch correction first to check if it is needed.
```
Is this comment already written earlier?

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,783 @@
+
+Now let's visualise the results again to see if the batch correction has worked. As before, we'll make one plot coloured by cluster and another coloured by batch (`Method`). We're hoping that the batches in that second plot will be more mixed together instead of forming separate groups like they did before batch correction.
+
+> <hands-on-title> Visualise the Results </hands-on-title>
```
I am not sure if you would like to add this suggestion: the user can also explore the results interactively by selecting yes to the Make an interactive plot? question in the Scanpy plot tool.

dianichj commented on topics/single-cell/tutorials/scrna_batch_correction/tutorial.md:
```diff
@@ -0,0 +1,783 @@
+
+# Checking the Clusters are Biologically Meaningful
+
+We've seen from our plots that the batch correction has mixed the different methods together, but this alone isn't enough to convince us that the batch correction has been successful. As always with single cell analysis, we also want to confirm that the clusters we've found are biologically meaningful. The Scanpy and Seurat pipelines will always present us with clusters, but it is up to us to mak ...
```
```suggestion
Our plots suggest that the batch correction has successfully brought the different methods together, but this alone is not enough to confirm that it has worked. As always in single cell analysis, we also need to verify that the clusters we have found are biologically meaningful. Scanpy and Seurat will always produce clusters, but it is up to us to evaluate whether those results actually make sense.
```

MarisaJL committed:
Apply suggestions from code review

Co-authored-by: Diana Chiang Jurado <122611454+dianichj@users.noreply.github.com>

MarisaJL committed:
Clarified sections on layers and added interactive visualisation suggestion

MarisaJL committed:
Enhance explanation of batch splitting in Seurat

Clarify the impact of splitting batches into layers in the Seurat pipeline.

MarisaJL committed:
Merge branch 'main' into batch_correct

pavanvidem reviewed (approved):
thanks @MarisaJL and @dianichj
